node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfraApplicationRedeployer.java

```diff
@@ -0,0 +1,117 @@
+import static java.util.logging.Level.WARNING;
+
+/**
+ * Performs on-demand redeployment of the {@link InfrastructureApplication}s, to minimise time between
+ * host provisioning for a deployment completing, and deployment of its application containers succeeding.
+ *
+ * @author jonmv
+ */
+public class InfraApplicationRedeployer extends AbstractComponent {
+
+    private static final Logger log = Logger.getLogger(InfraApplicationRedeployer.class.getName());
+
+    private final ExecutorService executor = Executors.newSingleThreadExecutor(new DaemonThreadFactory("infra-application-redeployer-"));
+    private final Set<InfrastructureApplication> readiedTypes = new ConcurrentSkipListSet<>();
+    private final InfraDeployer deployer;
+    private final Function<ApplicationId, Mutex> locks;
+    private final Supplier<NodeList> nodes;
+
+    @Inject
+    public InfraApplicationRedeployer(InfraDeployer deployer, NodeRepository nodes) {
+        this(deployer, nodes.applications()::lockMaintenance, nodes.nodes()::list);
+    }
+
+    InfraApplicationRedeployer(InfraDeployer deployer, Function<ApplicationId, Mutex> locks, Supplier<NodeList> nodes) {
+        this.deployer = deployer;
+        this.locks = locks;
+        this.nodes = nodes;
+    }
+
+    public void readied(NodeType type) {
+        readied(applicationOf(type));
+    }
+
+    private void readied(InfrastructureApplication application) {
+        if (application == null) return;
+        if (readiedTypes.add(application)) executor.execute(() -> checkAndRedeploy(application));
+    }
+
+    private void checkAndRedeploy(InfrastructureApplication application) {
+        log.log(INFO, () -> "Checking if " + application.name() + " should be redeployed");
+        if ( ! readiedTypes.remove(application)) return;
+        log.log(INFO, () -> "Trying to redeploy " + application.id() + " after completing provisioning of " + application.name());
+        try (Mutex lock = locks.apply(application.id())) {
+            if (application.nodeType().isHost() && nodes.get().state(State.ready).nodeType(application.nodeType()).isEmpty()) return;
+            log.log(INFO, () -> "Redeploying " + application.id() + " after completing provisioning of " + application.name());
+            try {
+                deployer.getDeployment(application.id()).ifPresent(Deployment::activate);
+                readied(childOf(application));
+            }
+            catch (RuntimeException e) {
+                log.log(WARNING, "Failed redeploying " + application.id() + ", will be retried by maintainer", e);
+            }
+        }
+        catch (UncheckedTimeoutException collision) {
+            readied(application);
+        }
+    }
+
+    private static InfrastructureApplication applicationOf(NodeType type) {
+        return switch (type) {
+            case host -> InfrastructureApplication.TENANT_HOST;
+            case confighost -> InfrastructureApplication.CONFIG_SERVER_HOST;
```

Comment:
Add config server and controller apps?

Reply:
Those are never set ready? We should only ever translate from host type nodes, I believe?

Reply:
I think cfg/ctl can be parked/failed, and then recycled with --> dirty --> ready --> reserved/active. I agree it's not on the common path.

Reply:
Handling it now.